### ISSUE TITLE
implement plan

### ISSUE THREAD

**Issue by danregima** [ReZorg/nn.nn#5]
<div class="UserMessageLayout-module__messageContainer__IOjwG"><span class="UserProblemStatementMessage-module__messageMeta__y5RUo">You asked</span><div style="--max-height: 500px;"><div class="ExpandableMarkdown-module__overflowing__D2ZUL ExpandableMarkdown-module__expanded__H8bpC"><div class="markdown-body MarkdownRenderer-module__container__sJ8jf ExpandableMarkdown-module__content__Rn4oS UserProblemStatementMessage-module__messageText__Jng0H" data-copilot-markdown="true" style="--MarkdownRenderer_streaming-fade-duration: 750ms;"><p>Implement the plan:</p>
<h1>Implementation Plan: Complete Features &amp; Functionalities in PLingua (<code node="[object Object]">lang/pli/</code>)</h1>
<h2>Current State Assessment</h2>
<p><strong>Already implemented (v1.0 core + v2.0 extensions):</strong></p>
<ul>
<li node="[object Object]">Core (<code node="[object Object]">nn.pli</code>): sigmoid/tanh/ReLU activations, linear layer, sequential container, softmax, MSE/NLL criteria, basic training loop, predict, XOR/binary/multiclass models</li>
<li node="[object Object]">Optimizers (<code node="[object Object]">momentum_adam_rmsprop.pli</code>): SGD+momentum, AdaGrad, RMSprop, Adam, AdamW, LR schedulers</li>
<li node="[object Object]">Normalization (<code node="[object Object]">batch_normalization.pli</code>), regularization (<code node="[object Object]">dropout.pli</code>), convolutional (<code node="[object Object]">convolutional.pli</code>), recurrent (<code node="[object Object]">recurrent.pli</code>), advanced backprop (<code node="[object Object]">backpropagation.pli</code>), visualization (<code node="[object Object]">visualization.pli</code>), integration tooling (<code node="[object Object]">integration.pli</code>)</li>
<li node="[object Object]">Tests (<code node="[object Object]">test_nn.pli</code>, 11 tests — core only), demos, examples, README</li>
</ul>
<p><strong>Identified gaps:</strong></p>
<ol>
<li node="[object Object]">README advertises BCE and Absolute Error criteria, but neither is implemented anywhere — docs/code mismatch</li>
<li node="[object Object]">Activation coverage is thin vs. the Lua reference (no LeakyReLU, ELU, GELU, PReLU, Softplus, HardTanh, LogSoftMax, etc.)</li>
<li node="[object Object]">Only one container (<code node="[object Object]">sequential</code>); Lua reference has Concat, Parallel, ConcatTable, and table-op modules</li>
<li node="[object Object]">Missing criteria beyond MSE/NLL: CrossEntropy (fused LogSoftMax+NLL), MarginCriterion, KL-divergence, Huber/SmoothL1</li>
<li node="[object Object]">No embedding/LookupTable or attention/transformer modules</li>
<li node="[object Object]">No Spiking Neural P-System (SN P) module — the natural membrane-computing formalism for neural nets, notably absent</li>
<li node="[object Object]">Weight initialization limited to uniform random (no Xavier/He)</li>
<li node="[object Object]">No serialization (save/load trained weights) despite an <code node="[object Object]">exporter</code> stub in <code node="[object Object]">integration.pli</code></li>
<li node="[object Object]">Test suite covers only v1.0 core — none of the 8 extension modules have tests</li>
<li node="[object Object]">Demos/examples don't exercise extension modules</li>
<li node="[object Object]">No CI/build wiring: <code node="[object Object]">.pli</code> files aren't referenced in <code node="[object Object]">CMakeLists.txt</code>, <code node="[object Object]">test/</code>, or any workflow; no syntax validation against the P-Lingua compiler</li>
<li node="[object Object]">README "Future Enhancements" checklist and "Limitations" section are stale (conv/recurrent/optimizers/dropout are now implemented but still listed as missing)</li>
</ol>
<h2>Phase 1: Fix Documentation–Code Mismatches</h2>
<ul>
<li node="[object Object]">Implement <code node="[object Object]">bce_criterion</code> and <code node="[object Object]">abs_criterion</code> modules in <code node="[object Object]">nn.pli</code> (forward loss + gradient rules, using the ×100 integer-scaling convention)</li>
<li node="[object Object]">Update README: mark completed "Future Enhancements" items, revise "Limitations" to reflect v2.0 modules, correct the feature list</li>
</ul>
<h2>Phase 2: Complete Core Module Parity</h2>
<ul>
<li node="[object Object]"><strong>Activations</strong> (new <code node="[object Object]">activations.pli</code>): LeakyReLU, PReLU, ELU, SELU, GELU, Softplus, HardTanh, HardSigmoid, LogSigmoid, LogSoftMax — each as a <code node="[object Object]">@module</code> with forward and derivative rules</li>
<li node="[object Object]"><strong>Criteria</strong> (extend <code node="[object Object]">nn.pli</code> or new <code node="[object Object]">criteria.pli</code>): CrossEntropy (fused), SmoothL1/Huber, Margin, KLDiv, weighted-class variants</li>
<li node="[object Object]"><strong>Containers</strong> (new <code node="[object Object]">containers.pli</code>): Concat (parallel branches, concatenated outputs), Parallel, ConcatTable/CAddTable-style table operations, Identity, Reshape/View analogs via object re-indexing</li>
<li node="[object Object]"><strong>Layers</strong>: LookupTable/embedding (index → weight-row objects), Bilinear, sparse linear</li>
<li node="[object Object]"><strong>Initialization</strong>: Xavier/Glorot and He initializers as parameterizable <code node="[object Object]">@initial</code> rules replacing hard-coded <code node="[object Object]">random(-50, 50)</code></li>
</ul>
<h2>Phase 3: Membrane-Computing-Native Features</h2>
<ul>
<li node="[object Object]"><strong>Spiking Neural P System module</strong> (new <code node="[object Object]">snp.pli</code>): spiking neurons with spike objects, firing/forgetting rules, delays — bridging this library to the canonical SN P formalism cited in the README references</li>
<li node="[object Object]"><strong>Attention module</strong> (new <code node="[object Object]">attention.pli</code>): scaled dot-product attention, multi-head attention, and a minimal transformer encoder block, building on existing softmax/linear/layer-norm modules</li>
<li node="[object Object]"><strong>Maximally-parallel batch processing</strong>: batch dimension encoded as an extra object index so multiple samples evolve concurrently (P-Systems' native strength)</li>
</ul>
<h2>Phase 4: Persistence &amp; Interop</h2>
<ul>
<li node="[object Object]">Complete <code node="[object Object]">exporter</code> in <code node="[object Object]">integration.pli</code>: weight snapshot/export rules and matching import (<code node="[object Object]">load_weights</code>) rules for round-tripping trained models</li>
<li node="[object Object]">Add a documented interchange format consistent with the other language ports (pl/rkt/scm) so weights can be shared across implementations</li>
</ul>
<h2>Phase 5: Tests, Demos, Examples</h2>
<ul>
<li node="[object Object]">Extend <code node="[object Object]">test_nn.pli</code> (or add <code node="[object Object]">test_extensions.pli</code>) with tests for: each optimizer's update rule, batch/layer norm train vs. eval, dropout masks, conv2d/pooling output shapes, RNN/LSTM/GRU gate evolution, gradient clipping, new activations/criteria, SN P module, attention</li>
<li node="[object Object]">Add demo sections for extension modules to <code node="[object Object]">demo.pli</code> and practical examples (e.g., LeNet-5 run, sequence prediction with LSTM, Adam vs. SGD comparison) to <code node="[object Object]">example.pli</code></li>
<li node="[object Object]">Update the README test-coverage list and demo/example lists accordingly</li>
</ul>
<h2>Phase 6: Tooling &amp; Validation</h2>
<ul>
<li node="[object Object]">Add a validation step that runs <code node="[object Object]">.pli</code> files through the P-Lingua 5 compiler/simulator (RGNC/plingua) for syntax checking; document how to invoke it</li>
<li node="[object Object]">Wire the pli test suite into the repository's existing test infrastructure (parallel to how <code node="[object Object]">lang/pl</code>, <code node="[object Object]">lang/rkt</code>, <code node="[object Object]">lang/scm</code> are handled) and add a CI job if workflows exist for the other ports</li>
<li node="[object Object]">Add a <code node="[object Object]">lang/pli</code> entry to top-level docs (<code node="[object Object]">README.md</code>, <code node="[object Object]">FORMAL_SPEC_INDEX.md</code> / mkdocs nav) if not already indexed</li>
</ul>
<h2>Suggested Order &amp; Dependencies</h2>
<ol>
<li node="[object Object]">Phase 1 (small, unblocks doc accuracy)</li>
<li node="[object Object]">Phase 2 (parity foundation; attention in Phase 3 depends on new activations/containers)</li>
<li node="[object Object]">Phase 3 (differentiating features)</li>
<li node="[object Object]">Phase 4 (needs stable module set)</li>
<li node="[object Object]">Phase 5 (tests written incrementally alongside Phases 2–4, consolidated here)</li>
<li node="[object Object]">Phase 6 (final validation and integration)</li>
</ol>
<h2>Risks / Considerations</h2>
<ul>
<li node="[object Object]">Integer ×100 scaling limits precision for GELU/softmax-heavy modules — may need a ×10000 fixed-point convention for new modules; document whichever is chosen</li>
<li node="[object Object]">No real PLingua interpreter is present in the repo, so "tests" are structural/rule-consistency checks unless Phase 6 tooling is added first; consider doing the compiler-validation harness early</li>
<li node="[object Object]">Keep the conceptual/educational framing: prefer clear rule-based expressions over performance tricks, consistent with the existing style</li>
</ul></div></div><button data-component="Button" type="button" aria-expanded="true" class="prc-Button-ButtonBase-9n-Xk ExpandableMarkdown-module__toggle__NBxAX" data-loading="false" data-no-visuals="true" data-size="small" data-variant="invisible"><span data-component="buttonContent" data-align="center" class="prc-Button-ButtonContent-Iohp5"><span data-component="text" class="prc-Button-Label-FWkx3">Show less</span></span></button></div></div>
